crates/biome_js_analyze/src/analyzers/nursery/no_unused_private_class_members.rs

```diff
@@ -0,0 +1,342 @@
+    ///
+    /// Private class members that are declared and not used anywhere in the code are most likely an error due to incomplete refactoring.
+    /// Such class members take up space in the code and can lead to confusion by readers.
+    ///
+    /// Source: https://eslint.org/docs/latest/rules/no-unused-private-class-members/
+    ///
+    /// ## Examples
+    ///
+    /// ### Invalid
+    ///
+    /// ```js,expect_diagnostic
+    /// class OnlyWrite {
+    ///  #usedOnlyInWrite = 5;
+    ///
+    ///  method() {
+    ///		this.#usedOnlyInWrite = 212;
+    ///  }
+    /// }
+    /// ```
+    ///
+    /// ```ts,expect_diagnostic
+    ///  class TsBioo {
+    ///    private unusedProperty = 5;
+    ///
+    ///    private unusedMethod() {
+
+    ///    };
+    ///  }
+    /// ```
+    ///
+    /// ## Valid
+    ///
+    /// ```js
+    /// class UsedMember {
+    ///  #usedMember = 42;
+    ///
+    ///  method() {
+    ///		return this.#usedMember;
+    ///  }
+    /// }
+    /// ```
+    ///
+    pub(crate) NoUnusedPrivateClassMembers {
+        version: "next",
+        name: "noUnusedPrivateClassMembers",
+        recommended: false,
+        fix_kind: FixKind::Unsafe,
+    }
+}
+
+declare_node_union! {
+    pub(crate) AnyMember = AnyJsClassMember | TsPropertyParameter
+}
+
+impl Rule for NoUnusedPrivateClassMembers {
+    type Query = Ast<JsClassDeclaration>;
+    type State = AnyMember;
+    type Signals = Vec<Self::State>;
+    type Options = ();
+
+    fn run(ctx: &RuleContext<Self>) -> Self::Signals {
+        let node = ctx.query();
+        let private_members: FxHashSet<AnyMember> = get_all_declared_private_members(node);
+
+        if private_members.is_empty() {
+            return vec![];
+        }
+
+        traverse_members_usage(node.syntax(), private_members)
+    }
+
+    fn diagnostic(_: &RuleContext<Self>, state: &Self::State) -> Option<RuleDiagnostic> {
+        Some(RuleDiagnostic::new(
+            rule_category!(),
+            state.property_range(),
+            markup! {
+                "This private class member is defined but never used."
+            },
+        ))
+    }
+
+    fn action(ctx: &RuleContext<Self>, state: &Self::State) -> Option<JsRuleAction> {
+        let mut mutation = ctx.root().begin();
+
+        mutation.remove_node(state.clone());
+
+        Some(JsRuleAction {
+            category: ActionCategory::QuickFix,
+            applicability: Applicability::MaybeIncorrect,
+            message: markup! { "Remove unused declaration." }.to_owned(),
+            mutation,
+        })
+    }
+}
+
+/// Check for private member usage
+/// if the member usage is found, we remove it from the hashmap
+fn traverse_members_usage(
+    syntax: &JsSyntaxNode,
+    mut private_members: FxHashSet<AnyMember>,
+) -> Vec<AnyMember> {
+    for node in syntax.descendants() {
+        if let Some(js_name) = AnyJsName::cast(node) {
+            private_members.retain(|private_member| {
```

Comment:
I think we should ensure that the private member usage is not in an inner class member.

```js
class A {
    #prop
    f() {
        class B {
            #prop
            g() { return this.#prop; }
        } 
    }
}
```

To do that you can check that we skip subtree that cast to `AnyJsControlFlowRoot` (note that we have to ignore ArrowFunctionExpression).

```rs
let mut it = syntax.preorder();
for node in it {
...
if AnyJsControlFlowRoot::can_cast(node.kind()) && !ArrowFunctionExpression::can_cast(node.kind()) {
    it.skip_subtree()
}
...
}
```

Unfortunately this requires to switch to `syntax.preorder()` to obtain an iterator with the `skip_subtree()` method